File: cpp/src/test_data_coupled.cpp
#include "test_data_coupled.h"
#include "test_data_utils.h"


TestDataCoupled::TestDataCoupled()
{
    std::random_device rd;
    mt = std::mt19937_64(rd());
    uniform_d = std::uniform_real_distribution<double>(0,1);

}
double TestDataCoupled::lam(double t, double mu, double gamma, std::vector<double>& St)
{
    if(gamma == 0)
        return mu;
    
    double g_sum = 0;

    for (double ti: St)
    {  
        if(ti < t)
          g_sum += g(t-ti);
    }    
    return mu + gamma * g_sum;
}

double TestDataCoupled::sample(int T, std::vector<double>& S_own, std::vector<double>& S_other, double t, double mu, double gamma, bool* add_t)
{
    double lam_upper = 0;
    for(double x = 0; x < T; x+=(((double)T)/10000.0))
    {
        double lam_val = lam(x, mu, gamma, S_other);

        if(lam_val > lam_upper)
            lam_upper = lam_val;
    }

    double u = uniform_d(mt);
    double t_new = t - 1.0 / lam_upper * log(u);
    double u2 = uniform_d(mt);
    *add_t = ((u2 <= (lam(t_new, mu, gamma, S_other) / lam_upper)) && (t_new < T));
    return t_new;
}


void TestDataCoupled::get_test_data(int T, double mu, double gamma_x, double gamma_y, double** Sx, int* len_Sx, double** Sy, int* len_Sy)
{
    double ty = 0.0, tx = 0.0, t_next = 0.0;
    std::vector<double>* Sy_vec = new vector<double>();
    std::vector<double>* Sx_vec = new vector<double>();
    bool get_X = true, get_Y = true;
    bool add_tx = false, add_ty = false;

    while(t_next < T)
    {
       
        if(get_X)
        {
            if(add_tx)
              Sx_vec->emplace_back(tx);
            tx = sample(T, *Sx_vec, *Sy_vec, tx, mu, gamma_y, &add_tx);
        }
        


        if(get_Y)
        {
            if(add_ty)
              Sy_vec->emplace_back(ty);
            ty = sample(T, *Sy_vec, *Sx_vec, ty, mu, gamma_x, &add_ty);
        }

        if(tx < ty)
        {
            get_X = true;
            get_Y = false;
            t_next = tx;
        }
        else if (ty < tx)
        {
            get_X = false;
            get_Y = true;
            t_next = ty;
        }
        else{
            cout << "ABORT: ty == tx" << endl;
            exit(-1);
        }
        t_next = double(int(t_next*3600) / 3600.0);
    }

    *Sy = &(*Sy_vec)[0];
    *len_Sy = (*Sy_vec).size();
    *Sx = &(*Sx_vec)[0];
    *len_Sx = (*Sx_vec).size();

    Sy_vec_heap = Sy_vec;
    Sx_vec_heap = Sx_vec;
}


TestDataCoupled::~TestDataCoupled()
{
    delete Sy_vec_heap;
    delete Sx_vec_heap;
}

File: py/test_data_coupled.py
"""Module test_data_coupled
   Generates and plots test data for TE estimations
   Users can influence each other.
"""





import numpy as np
import test_data_utils



def __lam(t, mu, gamma, St):
    if gamma == 0:
        return mu
    g_sum = sum([test_data_utils.g(t-ti) for ti in St if ti < t])
    return mu + gamma * g_sum


def __sample(T, S_own, S_other, t, mu, gamma): 
    lam_upper = max([__lam(x, mu, gamma, S_other) for x in np.linspace(0, t, num=100)])
    u = np.random.uniform()
    t = t - 1 / lam_upper * np.log(u)
    u2 = np.random.uniform()
    add_t = ((u2 <= (__lam(t, mu, gamma, S_other) / lam_upper)) and (t < T))
    return add_t, t



def get_test_data(T, mu, gamma_x, gamma_y):
    """Creates test data for users X and Y
    
    Arguments:
        T {int} -- time boundary in hours
        mu {float} -- background noise rate
        gamma_x {float} -- influence rate (How much does X influence Y?)
        gamma_y {float} -- influence rate (How much does Y influence X?)
        
    Returns:
        list -- timestamps Sx of activity of user X
        list -- timestamps Sy of activity of user Y
    """

    ty = tx = t_next = 0
    Sx = []
    Sy = []
    get_X = get_Y = True
    add_tx = add_ty = False
 
    while t_next < T:
               
        if get_X:
            if add_tx:
                Sx.append(tx)
            add_tx, tx = __sample(T, Sx, Sy, tx, mu, gamma_y)

        if get_Y:
            if add_ty:
                Sy.append(ty)
            add_ty, ty = __sample(T, Sy, Sx, ty, mu, gamma_x)

        if tx < ty:
            get_X = True
            get_Y = False
            t_next = tx
        elif ty < tx:
            get_X = False
            get_Y = True
            t_next = ty
        else:
            assert(False)

        print(t_next)

    return Sx, Sy




    

    
    

